File: backend/app/services/course_service.py
from sqlalchemy.orm import Session
from fastapi import HTTPException, status
from uuid import UUID
from datetime import datetime, timezone

from app.models.course import Course
from app.models.lesson import Lesson, LessonReviewStatus
from app.models.enrollment import Enrollment
from app.models.quiz import Quiz
from app.models.assignment import Assignment
from app.models.user import User, UserRole
from app.schemas.course import CourseCreate, CourseUpdate, CourseStatusUpdate, CourseStatus
from app.schemas.lesson import LessonReviewUpdate


def create_course(db: Session, data: CourseCreate, mentor: User) -> Course:
    course = Course(
        mentor_id=mentor.id,
        title=data.title,
        description=data.description,
        thumbnail=data.thumbnail,
        level=data.level,
        tags=data.tags,
        is_gated=data.is_gated,
        status=CourseStatus.pending,
    )
    db.add(course)
    db.commit()
    db.refresh(course)
    return course


def get_courses(db: Session, status: str = "approved", skip: int = 0, limit: int = 20):
    query = db.query(Course)
    if status:
        query = query.filter(Course.status == status)
    total = query.count()
    courses = query.offset(skip).limit(limit).all()
    return courses, total


def get_course_by_id(db: Session, course_id: UUID) -> Course:
    course = db.query(Course).filter(Course.id == course_id).first()
    if not course:
        raise HTTPException(status_code=404, detail="Course not found")
    return course


def update_course(db: Session, course_id: UUID, data: CourseUpdate, mentor: User) -> Course:
    course = get_course_by_id(db, course_id)
    if str(course.mentor_id) != str(mentor.id) and mentor.role != UserRole.admin:
        raise HTTPException(status_code=403, detail="Not your course")
    for field, value in data.model_dump(exclude_none=True).items():
        setattr(course, field, value)
    db.commit()
    db.refresh(course)
    return course


def approve_course(db: Session, course_id: UUID, data: CourseStatusUpdate, admin: User) -> Course:
    """Approve or reject a course with admin review notes."""
    course = get_course_by_id(db, course_id)
    
    # Validate that course is in pending status before approving/rejecting
    if course.status != CourseStatus.pending:
        raise HTTPException(
            status_code=400, 
            detail=f"Course is not pending approval. Current status: {course.status}"
        )
    
    course.status = data.status
    course.reviewed_by = admin.id
    course.reviewed_at = datetime.now(timezone.utc).replace(tzinfo=None)
    if data.reason:
        course.review_notes = data.reason
    db.commit()
    db.refresh(course)
    return course


def get_mentor_courses(db: Session, mentor_id: UUID):
    return db.query(Course).filter(Course.mentor_id == mentor_id).all()


def get_pending_courses(db: Session):
    return db.query(Course).filter(Course.status == CourseStatus.pending).all()


def get_pending_lessons(db: Session):
    """Get all lessons with pending review status across all courses."""
    return db.query(Lesson).filter(
        Lesson.review_status == LessonReviewStatus.pending
    ).all()


def review_lesson(db: Session, lesson_id: UUID, data: LessonReviewUpdate, admin: User) -> Lesson:
    """Admin review a lesson - approve or reject with feedback."""
    lesson = db.query(Lesson).filter(Lesson.id == lesson_id).first()
    if not lesson:
        raise HTTPException(status_code=404, detail="Lesson not found")
    
    lesson.review_status = LessonReviewStatus.approved if data.status == "approved" else LessonReviewStatus.rejected
    lesson.reviewed_by = admin.id
    lesson.reviewed_at = datetime.now(timezone.utc).replace(tzinfo=None)
    if data.feedback:
        lesson.review_feedback = data.feedback
    
    db.commit()
    db.refresh(lesson)
    return lesson


def get_course_lessons_for_review(db: Session, course_id: UUID) -> list:
    """Get all lessons of a course with their review status for admin review."""
    course = get_course_by_id(db, course_id)
    return course.lessons


def delete_course(db: Session, course_id: UUID, mentor: User) -> dict:
    """Delete a course and all related data. Only the creator or admin can delete.

    Explicitly deletes all dependent records (quizzes, assignments, enrollments)
    before deleting the course itself, providing defense-in-depth beyond ORM cascade.
    Returns a dict with info about what was deleted so the frontend
    can show appropriate feedback.
    """
    course = get_course_by_id(db, course_id)

    # Authorization: only the mentor who created the course (or admin) can delete
    if str(course.mentor_id) != str(mentor.id) and mentor.role != UserRole.admin:
        raise HTTPException(status_code=403, detail="Not your course")

    # Count records for feedback before deletion
    enrollment_count = db.query(Enrollment).filter(
        Enrollment.course_id == course_id
    ).count()
    lesson_count = db.query(Lesson).filter(
        Lesson.course_id == course_id
    ).count()
    quiz_count = db.query(Quiz).filter(
        Quiz.course_id == course_id
    ).count()
    assignment_count = db.query(Assignment).filter(
        Assignment.course_id == course_id
    ).count()

    # Explicitly delete all related records to avoid FK constraint issues.
    # The database has ON DELETE CASCADE, but SQLAlchemy's unit of work may
    # not respect it for relationships defined via backref without cascade.
    # This ensures clean deletion regardless of ORM relationship configuration.

    # Delete enrollments
    db.query(Enrollment).filter(
        Enrollment.course_id == course_id
    ).delete(synchronize_session=False)

    # Delete lessons (questions/quiz_attempts cascade from quizzes, submissions from assignments)
    db.query(Lesson).filter(
        Lesson.course_id == course_id
    ).delete(synchronize_session=False)

    # Delete quizzes (questions and quiz_attempts cascade from quiz model)
    db.query(Quiz).filter(
        Quiz.course_id == course_id
    ).delete(synchronize_session=False)

    # Delete assignments (submissions cascade from assignment model)
    db.query(Assignment).filter(
        Assignment.course_id == course_id
    ).delete(synchronize_session=False)

    # Now delete the course itself
    db.delete(course)
    db.commit()

    return {
        "deleted": True,
        "course_id": str(course_id),
        "enrollments_removed": enrollment_count,
        "lessons_removed": lesson_count,
        "quizzes_removed": quiz_count,
        "assignments_removed": assignment_count,
    }


def get_course_enrollment_count(db: Session, course_id: UUID) -> int:
    """Get the number of enrollments for a course."""
    return db.query(Enrollment).filter(
        Enrollment.course_id == course_id
    ).count()

File: frontend/src/services/courseService.ts
import api from "@/lib/axiosInstance";

export interface Course {
  id:             string;
  mentor_id:      string;
  title:          string;
  description:    string | null;
  thumbnail:      string | null;
  level:          "beginner" | "intermediate" | "advanced";
  tags:           string | null;
  status:         "draft" | "pending" | "approved" | "rejected";
  is_gated:       boolean;
  total_lessons:  number;
  total_enrolled: number;
  created_at:     string;
  mentor?:        { id: string; name: string };
}

export interface Lesson {
  id:             string;
  course_id:      string;
  title:          string;
  description:    string | null;
  youtube_url:    string;
  youtube_id:     string | null;
  duration:       number;
  order:          number;
  module:         string | null;
  is_free:        boolean;
  review_status:  "pending" | "approved" | "rejected";
  review_feedback?: string | null;
  created_at:     string;
}

export interface CreateCourseData {
  title:       string;
  description?: string;
  level:       string;
  tags?:       string;
  is_gated:    boolean;
}

export interface CreateLessonData {
  title:       string;
  description?: string;
  youtube_url: string;
  duration?:   number;
  order?:      number;
  module?:     string;
  is_free?:    boolean;
}

// Public — get all approved courses
export async function getCourses(): Promise<{ courses: Course[]; total: number }> {
  const res = await api.get("/courses");
  return res.data;
}

// Get single course
export async function getCourse(id: string): Promise<Course> {
  const res = await api.get(`/courses/${id}`);
  return res.data;
}

// Mentor — create course
export async function createCourse(data: CreateCourseData): Promise<Course> {
  const res = await api.post("/courses", data);
  return res.data;
}

// Mentor — get my courses
export async function getMyCourses(): Promise<Course[]> {
  const res = await api.get("/courses/my");
  return res.data;
}

// Admin — get pending courses
export async function getPendingCourses(): Promise<Course[]> {
  const res = await api.get("/courses/pending");
  return res.data;
}

// Admin — approve or reject course
export async function updateCourseStatus(
  id: string, status: string
): Promise<Course> {
  const res = await api.put(`/courses/${id}/approve`, { status });
  return res.data;
}

// Mentor — add lesson to course
export async function addLesson(
  courseId: string, data: CreateLessonData
): Promise<Lesson> {
  const res = await api.post(`/courses/${courseId}/lessons`, data);
  return res.data;
}

// Get lessons for a course
export async function getLessons(courseId: string): Promise<Lesson[]> {
  const res = await api.get(`/courses/${courseId}/lessons`);
  return res.data;
}

export interface PendingLesson {
  id:             string;
  course_id:      string;
  course_title:   string;
  title:          string;
  description:    string | null;
  youtube_url:    string;
  review_status:  string;
  created_at:     string;
}

// Admin — get all pending lessons
export async function getPendingLessons(): Promise<PendingLesson[]> {
  const res = await api.get("/courses/pending-lessons");
  return res.data;
}

// Admin — review (approve/reject) a lesson
export async function reviewLesson(
  lessonId: string, status: string, feedback?: string
): Promise<Lesson> {
  const res = await api.put(
    `/courses/lessons/${lessonId}/review`,
    { status, feedback }
  );
  return res.data;
}

// Mentor — delete a course
export interface DeleteCourseResult {
  deleted: boolean;
  course_id: string;
  enrollments_removed: number;
  lessons_removed: number;
  quizzes_removed: number;
  assignments_removed: number;
}

export async function deleteCourse(courseId: string): Promise<DeleteCourseResult> {
  const res = await api.delete(`/courses/${courseId}`);
  return res.data;
}

// Mentor — get enrollment count for a course (for deletion warnings)
export async function getCourseEnrollmentCount(courseId: string): Promise<{ course_id: string; enrollment_count: number }> {
  const res = await api.get(`/courses/${courseId}/enrollment-count`);
  return res.data;
}

// Mentor — delete a lesson
export interface DeleteLessonResult {
  deleted: boolean;
  lesson_id: string;
  course_id: string;
  enrollment_count: number;
}

export async function deleteLesson(courseId: string, lessonId: string): Promise<DeleteLessonResult> {
  const res = await api.delete(`/courses/${courseId}/lessons/${lessonId}`);
  return res.data;
}

// Mentor — get enrollment count for a lesson's parent course
export async function getLessonEnrollmentCount(courseId: string, lessonId: string): Promise<{ lesson_id: string; enrollment_count: number }> {
  const res = await api.get(`/courses/${courseId}/lessons/${lessonId}/enrollment-count`);
  return res.data;
}
